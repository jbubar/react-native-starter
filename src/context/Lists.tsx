import React, {createContext, useState, useCallback, useEffect} from 'react';
import firebase from 'firebase/app';

import {useCollectionData} from 'react-firebase-hooks/firestore';

export type Todo = {
  id: string;
  name: string;
};

const defaultState = {
  lists: [],
  selectedList: null,
};

const defaultActions = {
  addList: async (_newTodo: string) => {},
  selectList: (_todo: Todo) => {},
};

export const ListContext = createContext({
  isLoading: false,
  ...defaultState,
  ...defaultActions,
});

const ListProvider: React.FC = ({children}) => {
  const [selectedList, selectList] = useState({});

  const ListsCollection = firebase.firestore().collection('todoLists');

  const [lists, isLoading, error] = useCollectionData(ListsCollection, {
    idField: 'id',
  });
  useEffect(() => {
    if (lists) {
      selectList(lists[0]);
    }
  }, [lists]);

  const addList = useCallback(
    async ({name}) => {
      if (name.length) {
        try {
          return await ListsCollection.add({
            name,
          });
        } catch ({message}) {
          console.error(message);
        }
      }
    },
    [ListsCollection],
  );

  return (
    <ListContext.Provider
      value={{
        lists,
        isLoading,
        error: error?.message,
        addList,
        selectedList,
        selectList,
      }}
      children={children}
    />
  );
};

export const MockedProvider: React.FC = props => (
  <ListContext.Provider {...props} />
);

export default ListProvider;
