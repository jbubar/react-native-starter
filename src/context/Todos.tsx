import React, {
  createContext,
  useState,
  useCallback,
  useContext,
  useMemo,
} from 'react';
import firebase from 'firebase/app';

import {useCollectionData} from 'react-firebase-hooks/firestore';
import {ListContext} from './Lists';

export type Todo = {
  id: string;
  description: string;
  done: boolean;
  date?: Date;
};

const defaultState = {
  todos: [],
  selectedTodo: null,
};

const defaultActions = {
  addTodo: async (_newTodo: string) => {},
  toggleDone: async (_todo: Todo) => {},
  saveTodo: async (_todo: Todo) => {},
  deleteTodo: async (_todo: Todo) => {},
  setSelected: (_todo: Todo) => {},
};

export const TodoContext = createContext({
  isLoading: false,
  ...defaultState,
  ...defaultActions,
});

const TodoProvider: React.FC = ({children}) => {
  const {selectedList} = useContext(ListContext);

  const [selectedTodo, setSelected] = useState(null);

  const TodosCollection = useMemo(
    () =>
      firebase
        .firestore()
        .collection('todoLists')
        .doc(selectedList.id)
        .collection('todos'),
    [selectedList],
  );

  const [todos, isLoading, error] = useCollectionData(TodosCollection, {
    idField: 'id',
  });

  const addTodo = useCallback(
    async newTodo => {
      if (newTodo.length) {
        try {
          await TodosCollection.add({
            description: newTodo,
            done: false,
          });
        } catch ({message}) {
          console.error(message);
        }
      }
    },
    [TodosCollection],
  );

  const toggleDone = useCallback(
    async ({id, done}: Todo) => {
      try {
        await TodosCollection.doc(id).update({done: !done});
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  const saveTodo = useCallback(
    async ({id, description, done, date}: Todo) => {
      try {
        await TodosCollection.doc(id).update({
          description,
          done,
          date: firebase.firestore.Timestamp.fromDate(date),
        });
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  const deleteTodo = useCallback(
    async ({id}: Todo) => {
      try {
        await TodosCollection.doc(id).delete();
      } catch (e) {
        console.error(e);
      }
    },
    [TodosCollection],
  );

  return (
    <TodoContext.Provider
      value={{
        todos,
        isLoading,
        error: error?.message,
        addTodo,
        toggleDone,
        selectedTodo,
        setSelected,
        saveTodo,
        deleteTodo,
      }}
      children={children}
    />
  );
};

export const MockedProvider: React.FC = props => (
  <TodoContext.Provider {...props} />
);

export default TodoProvider;
