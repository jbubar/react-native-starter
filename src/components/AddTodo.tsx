import React, {useCallback, useContext, useState} from 'react';
import {StyleSheet} from 'react-native';

import {Input, Layout, Button} from '@ui-kitten/components';

import {TodoContext} from '../context/Todos';

const styles = StyleSheet.create({
  inputRow: {flexDirection: 'row', padding: 10},
  messageInput: {flex: 1},
  addButton: {marginLeft: 10},
});

export default () => {
  const {addTodo} = useContext(TodoContext);
  const [newTodo, setNewTodo] = useState('');

  const submit = useCallback(async () => {
    await addTodo(newTodo);
    setNewTodo('');
  }, [addTodo, newTodo]);

  return (
    <Layout style={styles.inputRow}>
      <Input
        style={styles.messageInput}
        size="large"
        placeholder="Add a todo..."
        value={newTodo}
        onChangeText={setNewTodo}
        onSubmitEditing={submit}
        testID="todoInput"
      />
      <Button onPress={submit} style={styles.addButton} testID="addButton">
        +
      </Button>
    </Layout>
  );
};
