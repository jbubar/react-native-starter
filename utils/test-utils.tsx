// test-utils.js

import React, {ReactElement} from 'react';
import {SafeAreaView, StyleSheet} from 'react-native';

import * as eva from '@eva-design/eva';
import {EvaIconsPack} from '@ui-kitten/eva-icons';
import {ApplicationProvider, IconRegistry} from '@ui-kitten/components';

import {render, RenderOptions} from '@testing-library/react-native';

const styles = StyleSheet.create({
  safeArea: {flex: 1},
});

const AllTheProviders: React.FC = ({children}) => {
  return (
    <>
      <IconRegistry icons={EvaIconsPack} />
      <ApplicationProvider {...eva} theme={eva.light}>
        <SafeAreaView style={styles.safeArea}>{children}</SafeAreaView>
      </ApplicationProvider>
    </>
  );
};

const customRender = (ui: ReactElement, options: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
