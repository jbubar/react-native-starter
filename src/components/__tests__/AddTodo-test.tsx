/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';

import {render, fireEvent, waitFor} from 'test-utils';

import AddTodo from '../AddTodo';
import {MockedProvider} from '../../context/Todos';

jest.mock('react-native-config', () => ({
  Config: {
    SINGLE_LIST_ID: '1',
  },
}));

describe('AddTodo', () => {
  let wrapper, addTodo;

  beforeAll(() => {
    jest.resetAllMocks();
  });

  beforeEach(() => {
    addTodo = jest.fn(todo => Promise.resolve());
    wrapper = render(
      <MockedProvider value={{addTodo}}>
        <AddTodo />
      </MockedProvider>,
    );
  });

  test('Snapshot', () => {
    const {toJSON} = wrapper;
    expect(toJSON()).toMatchSnapshot();
  });

  test('Add Todo With Button', async () => {
    const {getByTestId} = wrapper;
    fireEvent.changeText(getByTestId('todoInput'), 'new todo');
    fireEvent.press(getByTestId('addButton'));

    await waitFor(() => {
      expect(addTodo).toHaveBeenCalledWith('new todo');
    });
  });

  test('Add Todo With Return/Enter', async () => {
    const {getByTestId} = wrapper;

    const todoInput = getByTestId('todoInput');
    fireEvent.changeText(todoInput, 'new todo');
    fireEvent(todoInput, 'submitEditing');

    await waitFor(() => {
      expect(addTodo).toHaveBeenCalledWith('new todo');
    });
  });
});
