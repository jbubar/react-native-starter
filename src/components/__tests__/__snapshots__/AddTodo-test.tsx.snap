// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo Snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      appearance="default"
      level="1"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "flexDirection": "row",
            "padding": 10,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#F7F9FC",
                "borderColor": "#E4E9F2",
                "borderRadius": 4,
                "borderWidth": 1,
                "minHeight": 48,
                "paddingHorizontal": 8,
                "paddingVertical": 11,
              },
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            allowFontScaling={true}
            appearance="default"
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Add a todo..."
            placeholderTextColor="#8F9BB3"
            rejectResponderTermination={true}
            size="large"
            status="basic"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "normal",
                  "marginHorizontal": 8,
                },
                Object {
                  "flexBasis": "auto",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                null,
                undefined,
              ]
            }
            testID="todoInput"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3366FF",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginLeft": 10,
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 12,
          }
        }
        testID="addButton"
      >
        <Text
          appearance="default"
          category="p1"
          style={
            Array [
              Object {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "bold",
                "marginHorizontal": 10,
              },
            ]
          }
        >
          +
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
